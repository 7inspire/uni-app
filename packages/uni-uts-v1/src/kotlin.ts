import os from 'os'
import fs from 'fs-extra'
import path from 'path'
import AdmZip from 'adm-zip'
import { sync } from 'fast-glob'
import { isArray } from '@vue/shared'
import {
  isInHBuilderX,
  normalizePath,
  parseJson,
  resolveSourceMapPath,
} from './shared'
import {
  CompilerServer,
  genUTSPlatformResource,
  getUtsCompiler,
  getCompilerServer,
  moveRootIndexSourceMap,
  resolveAndroidDir,
  resolvePackage,
  resolveUTSPlatformFile,
  resolveUTSSourceMapPath,
} from './utils'
import { Module } from '../types/types'
import { UtsResult } from '@dcloudio/uts'

interface KotlinCompilerServer extends CompilerServer {
  getKotlincHome(): string
  getDefaultJar(): string[]
  compile(
    options: { kotlinc: string[]; d8: string[] },
    projectPath: string
  ): Promise<boolean>
  checkDependencies?: (
    configJsonPath: string
  ) => Promise<{ code: number; msg: string; data: string[] }>
  checkRResources?: (resDir: string) => Promise<{
    code: number
    msg: string
    data: { jarPath: string; uniModuleName: string }
  }>
}

export function createKotlinResolveTypeReferenceName(
  _namespace: string,
  _ast: Module
) {
  return (name: string) => name
}

function parseKotlinPackage(filename: string) {
  const res = resolvePackage(filename)
  if (!res) {
    return { package: '' }
  }
  return {
    package: 'uts.sdk.' + (res.is_uni_modules ? 'modules.' : '') + res.name,
  }
}

export async function runKotlinProd(filename: string) {
  // 文件有可能是 app-ios 里边的，因为编译到 android 时，为了保证不报错，可能会去读取 ios 下的 uts
  if (filename.includes('app-ios')) {
    return
  }
  await compile(filename)
  genUTSPlatformResource(filename, {
    inputDir: process.env.UNI_INPUT_DIR,
    outputDir: process.env.UNI_OUTPUT_DIR,
    platform: 'app-android',
    extname: '.kt',
  })
}

export type RunKotlinDevResult = UtsResult & {
  type: 'kotlin'
  changed: string[]
}

export async function runKotlinDev(
  filename: string
): Promise<RunKotlinDevResult | undefined> {
  // 文件有可能是 app-ios 里边的，因为编译到 android 时，为了保证不报错，可能会去读取 ios 下的 uts
  if (filename.includes('app-ios')) {
    return
  }
  const result = (await compile(filename)) as RunKotlinDevResult

  result.type = 'kotlin'
  result.changed = []

  const kotlinFile = resolveUTSPlatformFile(filename, {
    inputDir: process.env.UNI_INPUT_DIR,
    outputDir: process.env.UNI_OUTPUT_DIR,
    platform: 'app-android',
    extname: '.kt',
  })
  // 开发模式下，需要生成 dex
  if (fs.existsSync(kotlinFile)) {
    const compilerServer = getCompilerServer<KotlinCompilerServer>(
      'uniapp-runextension'
    )
    if (!compilerServer) {
      throw `项目使用了uts插件，正在安装 uts Android 运行扩展...`
    }
    const {
      getDefaultJar,
      getKotlincHome,
      compile,
      checkDependencies,
      checkRResources,
    } = compilerServer
    let deps: string[] = []
    if (checkDependencies) {
      deps = await checkDeps(filename, checkDependencies)
    }
    let resDeps: string[] = []
    if (checkRResources) {
      resDeps = await checkRes(filename, checkRResources)
    }
    // time = Date.now()
    const jarFile = resolveJarPath(kotlinFile)
    const options = {
      kotlinc: resolveKotlincArgs(
        kotlinFile,
        getKotlincHome(),
        getDefaultJar()
          .concat(resolveLibs(filename))
          .concat(deps)
          .concat(resDeps)
      ),
      d8: resolveD8Args(jarFile),
      sourceRoot: process.env.UNI_INPUT_DIR,
      sourceMapPath: resolveSourceMapFile(
        process.env.UNI_OUTPUT_DIR,
        kotlinFile
      ),
    }
    const res = await compile(options, process.env.UNI_INPUT_DIR)
    // console.log('dex compile time: ' + (Date.now() - time) + 'ms')
    if (res) {
      try {
        fs.unlinkSync(jarFile)
        // 短期内先不删除，方便排查问题
        // fs.unlinkSync(kotlinFile)
      } catch (e) {}
      const dexFile = resolveDexFile(jarFile)
      if (fs.existsSync(dexFile)) {
        result.changed = [
          normalizePath(path.relative(process.env.UNI_OUTPUT_DIR, dexFile)),
        ]
      }
    }
    // else {
    //   throw `${normalizePath(
    //     path.relative(process.env.UNI_INPUT_DIR, filename)
    //   )} 编译失败`
    // }
    if (process.env.HX_USE_BASE_TYPE === 'standard') {
      if (!isSupportStandardPlayground(filename)) {
        const pkg = resolvePackage(filename)
        if (pkg) {
          console.warn(
            `uts插件[${pkg.id}]依赖的原生配置或三方SDK在运行至标准基座时不能生效，如需正常调用请使用自定义基座`
          )
        }
      }
    }
  }
  return result
}

function checkDeps(
  filename: string,
  checkDependencies: (
    configJsonFile: string
  ) => Promise<{ code: number; msg: string; data: string[] }>
) {
  const configJsonFile = resolveConfigJsonFile(filename)
  if (configJsonFile && hasDeps(configJsonFile)) {
    return checkDependencies(configJsonFile).then(({ code, msg, data }) => {
      if (code !== 0) {
        console.error(msg)
        return []
      }
      return data
    })
  }
  return Promise.resolve([])
}

function hasDeps(configJsonFile: string) {
  const deps =
    parseJson(fs.readFileSync(configJsonFile, 'utf8')).dependencies || []
  if (isArray(deps) && deps.length) {
    return true
  }
  return false
}

function checkRes(
  filename: string,
  checkRResources: (resDir: string) => Promise<{
    code: number
    msg: string
    data: { jarPath: string; uniModuleName: string }
  }>
) {
  const resDir = resolveResDir(filename)
  if (resDir) {
    return checkRResources(resDir).then(({ code, msg, data }) => {
      if (code !== 0) {
        console.error(msg)
        return []
      }
      return [data.jarPath]
    })
  }
  return Promise.resolve([])
}

function resolveResDir(filename: string) {
  const resDir = path.resolve(resolveAndroidDir(filename), 'res')
  if (fs.existsSync(resDir)) {
    return resDir
  }
}

function resolveAndroidResourceClass(filename: string) {
  const resDir = resolveResDir(filename)
  if (resDir && fs.readdirSync(resDir).length) {
    const pkg = resolveAndroidManifestPackage(filename)
    if (pkg) {
      return pkg + '.R'
    }
  }
}

const packageRe = /\s+package="(.*)"/
function resolveAndroidManifestPackage(filename: string) {
  const manifestXmlPath = path.resolve(
    resolveAndroidDir(filename),
    'AndroidManifest.xml'
  )
  if (fs.existsSync(manifestXmlPath)) {
    const matches = fs.readFileSync(manifestXmlPath, 'utf8').match(packageRe)
    if (matches && matches[1]) {
      return matches[1]
    }
  }
}

function resolveConfigJsonFile(filename: string) {
  const configJsonFile = path.resolve(
    resolveAndroidDir(filename),
    'config.json'
  )
  if (fs.existsSync(configJsonFile)) {
    return configJsonFile
  }
}

function resolveSourceMapFile(outputDir: string, kotlinFile: string) {
  return (
    path.resolve(resolveSourceMapPath(), path.relative(outputDir, kotlinFile)) +
    '.map'
  )
}

const DEFAULT_IMPORTS = [
  'kotlinx.coroutines.async',
  'kotlinx.coroutines.CoroutineScope',
  'kotlinx.coroutines.Deferred',
  'kotlinx.coroutines.Dispatchers',
  'io.dcloud.uts.*',
]

async function compile(filename: string) {
  if (!process.env.UNI_HBUILDERX_PLUGINS) {
    throw 'process.env.UNI_HBUILDERX_PLUGINS is not found'
  }
  const { bundle, UtsTarget } = getUtsCompiler()
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  // let time = Date.now()
  const imports = [...DEFAULT_IMPORTS]
  const rClass = resolveAndroidResourceClass(filename)
  if (rClass) {
    imports.push(rClass)
  }
  const result = await bundle(UtsTarget.KOTLIN, {
    input: {
      root: inputDir,
      filename,
    },
    output: {
      isPlugin: true,
      outDir: outputDir,
      package: parseKotlinPackage(filename).package,
      sourceMap: resolveUTSSourceMapPath(filename),
      extname: 'kt',
      imports,
      logFilename: true,
      noColor: isInHBuilderX(),
    },
  })
  moveRootIndexSourceMap(filename, {
    inputDir: process.env.UNI_INPUT_DIR,
    outputDir: process.env.UNI_OUTPUT_DIR,
    platform: 'app-android',
    extname: '.kt',
  })
  return result
}

function resolveKotlincArgs(filename: string, kotlinc: string, jars: string[]) {
  return [
    filename,
    '-cp',
    resolveClassPath(jars),
    '-d',
    resolveJarPath(filename),
    '-kotlin-home',
    kotlinc,
  ]
}

function resolveD8Args(filename: string) {
  return [
    filename,
    '--no-desugaring',
    '--min-api',
    '19',
    '--output',
    resolveDexPath(filename),
  ]
}

function resolveLibs(filename: string) {
  const libsPath = path.resolve(resolveAndroidDir(filename), 'libs')
  const libs: string[] = []
  if (fs.existsSync(libsPath)) {
    libs.push(...sync('*.jar', { cwd: libsPath, absolute: true }))
    const zips = sync('*.aar', { cwd: libsPath })
    zips.forEach((name) => {
      const outputPath = resolveAndroidArchiveOutputPath(name)
      if (!fs.existsSync(outputPath)) {
        // 解压
        const zip = new AdmZip(path.resolve(libsPath, name))
        zip.extractAllTo(outputPath, true)
      }
      libs.push(
        ...sync('**/*.jar', {
          cwd: outputPath,
          absolute: true,
        })
      )
    })
  }
  return libs
}

function resolveAndroidArchiveOutputPath(aar?: string) {
  return path.resolve(
    process.env.UNI_OUTPUT_DIR,
    '../.uts/aar',
    aar ? aar.replace('.aar', '') : ''
  )
}
function resolveDexFile(jarFile: string) {
  return normalizePath(path.resolve(path.dirname(jarFile), 'classes.dex'))
}

function resolveDexPath(filename: string) {
  return path.dirname(filename)
}

function resolveJarPath(filename: string) {
  return filename.replace(path.extname(filename), '.jar')
}

function resolveClassPath(jars: string[]) {
  return jars.join(os.platform() === 'win32' ? ';' : ':')
}

const checkFiles = ['AndroidManifest.xml']
const checkDirs = ['libs', 'assets', 'res']
/**
 * 当前插件是否支持标准基座运行
 * @param filename
 * @returns
 */
function isSupportStandardPlayground(filename: string) {
  const baseDir = resolveAndroidDir(filename)
  if (checkFiles.find((file) => fs.existsSync(path.resolve(baseDir, file)))) {
    return false
  }
  if (
    checkDirs.find((dir) => {
      const absDir = path.resolve(baseDir, dir)
      return fs.existsSync(absDir) && fs.readdirSync(absDir).length
    })
  ) {
    return false
  }
  const configJsonFile = resolveConfigJsonFile(filename)
  if (configJsonFile && hasDeps(configJsonFile)) {
    return false
  }
  return true
}
