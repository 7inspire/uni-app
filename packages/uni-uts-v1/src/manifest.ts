import md5 from 'md5-file'
import glob from 'fast-glob'
import { normalizePath } from './shared'
export async function genManifest(pluginDir: string) {
  const files = await glob(normalizePath(pluginDir) + '/**/*')
  await Promise.all(files.map((file) => md5(file)))
}
