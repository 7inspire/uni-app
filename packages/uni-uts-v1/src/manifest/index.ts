import {
  checkManifest,
  hasCustomResources,
  isCustomResources,
  resolveManifestJson,
  resolvePluginFiles,
} from './manifest'
import {
  APP_PLATFORM,
  CheckOptions,
  CheckResult,
  customResourceChangedTips,
  customResourceTips,
} from './utils'

export { genManifestFile } from './manifest'

export { restoreDex, storeDex } from './dex'
export { restoreSourceMap, storeSourceMap } from './sourceMap'

interface PlatformOptions {
  customRes: string[]
}

const ANDROID_CUSTOM_RES = [
  'app-android/assets/',
  'app-android/libs/',
  'app-android/res/',
  'app-android/AndroidManifest.xml',
  'app-android/config.json',
]

const IOS_CUSTOM_RES = [
  'app-ios/Frameworks/',
  'app-ios/Resources/',
  'app-ios/Info.plist',
  'app-ios/config.json',
]

export function checkKotlinCompile(
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  return checkCompile('app-android', playground, options)
}

export function checkSwiftCompile(
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  return checkCompile('app-ios', playground, options)
}

export function checkCompile(
  platform: APP_PLATFORM,
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  const platformOptions: PlatformOptions = {
    customRes: platform === 'app-android' ? ANDROID_CUSTOM_RES : IOS_CUSTOM_RES,
  }
  if (playground === 'standard') {
    return checkWithPlayground(platform, 'standard', options, platformOptions)
  }
  return checkWithPlayground(platform, 'custom', options, platformOptions)
}

async function checkWithPlayground(
  platform: APP_PLATFORM,
  type: typeof process.env.HX_USE_BASE_TYPE,
  {
    id,
    env,
    cacheDir,
    pluginDir,
    pluginRelativeDir,
    is_uni_modules,
  }: CheckOptions,
  { customRes }: PlatformOptions
): Promise<CheckResult> {
  // 第一步：获取所有文件列表
  const files = await resolvePluginFiles(platform, pluginDir, is_uni_modules)
  let tips = ''
  // 标准基座检查是否包含原生资源/配置
  if (type === 'standard' && hasCustomResources(files, customRes)) {
    tips = customResourceTips(id)
  }
  // 第二步：获取当前插件缓存文件信息
  const manifest = resolveManifestJson(platform, pluginRelativeDir, cacheDir)
  if (!manifest) {
    return { expired: true, tips, files }
  }
  // 第四步：检查文件变更
  const res = await checkManifest(manifest, { env, files, pluginDir })
  // 自定义基座检查原生资源/配置是否发生变化
  if (type === 'custom' && typeof res === 'string') {
    if (isCustomResources(res, customRes)) {
      tips = customResourceChangedTips(id)
    }
  }
  return {
    expired: res !== true,
    tips,
    files,
  }
}

export function initCheckOptionsEnv(): CheckOptions['env'] {
  return {
    compilerVersion: require('../../package.json').version,
  }
}
