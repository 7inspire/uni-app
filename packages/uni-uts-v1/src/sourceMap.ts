import { existsSync, readFileSync, statSync } from 'fs'
import { basename, dirname, join, relative } from 'path'
import {
  BasicSourceMapConsumer,
  IndexedSourceMapConsumer,
  NullableMappedPosition,
  NullablePosition,
  SourceMapConsumer,
} from 'source-map'

const EXTNAME = {
  kotlin: '.kt',
  swift: '.swift',
}

const PLATFORM_DIR = {
  kotlin: 'app-android',
  swift: 'app-ios',
}

export function resolveUtsPluginSourceMapFile(
  target: 'kotlin' | 'swift',
  filename: string,
  inputDir: string,
  outputDir: string
) {
  inputDir = normalizePath(inputDir)
  outputDir = normalizePath(outputDir)
  filename = normalizePath(filename)
  const pluginDir = resolvePluginDir(inputDir, outputDir, filename)
  if (!pluginDir) {
    throw `plugin dir not found`
  }
  const is_uni_modules = basename(dirname(pluginDir)) === 'uni_modules'
  const sourceMapFile = join(
    join(outputDir, '../.sourcemap/app'),
    relative(inputDir, pluginDir),
    is_uni_modules ? 'utssdk' : '',
    PLATFORM_DIR[target],
    `index${EXTNAME[target]}.map`
  )
  if (!existsSync(sourceMapFile)) {
    throw `${sourceMapFile} not found`
  }
  return sourceMapFile
}

function resolvePluginDir(
  inputDir: string,
  outputDir: string,
  filename: string
) {
  // 目标文件是编译后 kt 或 swift
  if (filename.startsWith(outputDir)) {
    const relativePath = relative(outputDir, filename)
    // uni_modules/test-uts
    if (relativePath.startsWith('uni_modules')) {
      return join(inputDir, join(relativePath, '../../..'))
    }
    // utssdk/test-uts
    return join(inputDir, join(relativePath, '../..'))
  } else if (filename.startsWith(inputDir)) {
    let parent = dirname(filename)
    const utssdkDir = normalizePath(join(inputDir, 'utssdk'))
    const uniModulesDir = normalizePath(join(inputDir, 'uni_modules'))
    while (parent) {
      const dir = dirname(parent)
      if (parent === dir) {
        // windows 上边会剩下一个盘符
        return
      }
      if (dir === utssdkDir || dir === uniModulesDir) {
        return parent
      }
      parent = dir
    }
    throw `${filename} is not a uts plugin file`
  } else {
    throw `${filename} is not in ${inputDir} or ${outputDir}`
  }
}

enum BIAS {
  GREATEST_LOWER_BOUND = 1,
  LEAST_UPPER_BOUND = 2,
}

interface PositionFor {
  sourceMapFile: string
  filename: string
  line: number
  column: number
  withSourceContent?: boolean
}

const consumers: Record<
  string,
  { time: number; consumer: BasicSourceMapConsumer | IndexedSourceMapConsumer }
> = {}

/**
 * 解析源码文件，目前 uts 的 sourcemap 存储的都是相对目录
 * @param consumer
 * @param filename
 * @returns
 */
function resolveSource(
  consumer: BasicSourceMapConsumer | IndexedSourceMapConsumer,
  filename: string
) {
  filename = normalizePath(filename)
  return (
    consumer.sources.find((source) => filename.endsWith(source)) || filename
  )
}

/**
 * 根据源码文件名、行号、列号，返回生成后文件、行号、列号（根据 uts 文件返回 kt|swift 文件）
 * @param originalPosition
 * @returns
 */
export function generatedPositionFor({
  sourceMapFile,
  filename,
  line,
  column,
  outputDir,
}: PositionFor & { outputDir?: string }): Promise<
  NullablePosition & { source: string | null }
> {
  return resolveSourceMapConsumer(sourceMapFile).then((consumer) => {
    const res = consumer.generatedPositionFor({
      source: resolveSource(consumer, filename),
      line,
      column,
      bias: column === 0 ? BIAS.LEAST_UPPER_BOUND : BIAS.GREATEST_LOWER_BOUND,
    })
    let source = null
    if (outputDir) {
      // 根据 sourceMapFile 和 outputDir，计算出生成后的文件路径
      source = join(
        outputDir,
        relative(join(outputDir, '../.sourcemap/app'), sourceMapFile)
      ).replace('.map', '')
    }
    return Object.assign(res, { source })
  })
}

/**
 * 根据生成后的文件名、行号、列号，返回源码文件、行号、列号（根据 kt|swift 文件返回 uts 文件）
 * @param generatedPosition
 * @returns
 */
export function originalPositionFor(
  generatedPosition: Omit<PositionFor, 'filename'> & { inputDir?: string }
): Promise<NullableMappedPosition & { sourceContent?: string }> {
  return resolveSourceMapConsumer(generatedPosition.sourceMapFile).then(
    (consumer) => {
      const res = consumer.originalPositionFor({
        line: generatedPosition.line,
        column: generatedPosition.column,
        bias:
          generatedPosition.column === 0
            ? BIAS.LEAST_UPPER_BOUND
            : BIAS.GREATEST_LOWER_BOUND,
      })
      if (
        generatedPosition.withSourceContent &&
        res.source &&
        res.line &&
        res.column
      ) {
        return Object.assign(res, {
          sourceContent: consumer.sourceContentFor(res.source, true),
        })
      }
      if (res.source && generatedPosition.inputDir) {
        res.source = join(generatedPosition.inputDir, res.source)
      }
      return res
    }
  )
}

async function resolveSourceMapConsumer(sourceMapFile: string) {
  const stats = statSync(sourceMapFile)
  if (!stats.isFile()) {
    throw `${sourceMapFile} is not a file`
  }
  const cache = consumers[sourceMapFile]
  if (!cache || cache.time !== stats.mtimeMs) {
    consumers[sourceMapFile] = {
      time: stats.mtimeMs,
      consumer: await new SourceMapConsumer(
        readFileSync(sourceMapFile, 'utf8')
      ),
    }
  }
  return consumers[sourceMapFile].consumer
}

function normalizePath(path: string) {
  return path.replace(/\\/g, '/')
}
