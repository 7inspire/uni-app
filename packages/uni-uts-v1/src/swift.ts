import fs from 'fs-extra'
import path, { join } from 'path'
import { capitalize } from '@vue/shared'
import {
  genUTSPlatformResource,
  getCompilerServer,
  getUtsCompiler,
  moveRootIndexSourceMap,
  resolveIOSDir,
  resolvePackage,
  resolveUTSPlatformFile,
  resolveUTSSourceMapPath,
  ToSwiftOptions,
} from './utils'
import { isInHBuilderX, parseJson } from './shared'
import { UtsResult } from '@dcloudio/uts'

function parseSwiftPackage(filename: string) {
  const res = resolvePackage(filename)
  if (!res) {
    return {
      namespace: '',
    }
  }
  const namespace =
    'UTSSDK' + (res.is_uni_modules ? 'Modules' : '') + capitalize(res.name)
  return {
    namespace,
  }
}

export async function runSwiftProd(filename: string) {
  // 文件有可能是 app-android 里边的，因为编译到 ios 时，为了保证不报错，可能会去读取 android 下的 uts
  if (filename.includes('app-android')) {
    return
  }
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  await compile(filename, { inputDir, outputDir, sourceMap: true })
  genUTSPlatformResource(filename, {
    inputDir,
    outputDir,
    platform: 'app-ios',
    extname: '.swift',
  })
}

export type RunSwiftDevResult = UtsResult & {
  type: 'swift'
  code: number
  msg: string
  changed: string[]
}

let isEnvReady = true
export async function runSwiftDev(filename: string) {
  // 文件有可能是 app-android 里边的，因为编译到 ios 时，为了保证不报错，可能会去读取 android 下的 uts
  if (filename.includes('app-android')) {
    return
  }
  if (!isEnvReady) {
    console.error(`已跳过uts插件[${resolvePackage(filename)?.id}]的编译`)
    return
  }
  const compilerServer = getCompilerServer<SwiftCompilerServer>(
    'uts-development-ios'
  )
  if (!compilerServer) {
    throw `项目使用了uts插件，正在安装 uts iOS 运行扩展...`
  }
  if (compilerServer.checkEnv) {
    const { code, msg } = compilerServer.checkEnv()
    if (code) {
      isEnvReady = false
      console.error(msg)
      return
    }
  }
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  const result = (await compile(filename, {
    inputDir,
    outputDir,
    sourceMap: true,
  })) as RunSwiftDevResult

  result.type = 'swift'

  const swiftFile = resolveUTSPlatformFile(filename, {
    inputDir,
    outputDir,
    platform: 'app-ios',
    extname: '.swift',
  })
  result.changed = []
  // 开发模式下，需要生成 framework
  if (fs.existsSync(swiftFile)) {
    let projectPath = inputDir
    const isCli = isCliProject(projectPath)
    if (isCli) {
      projectPath = path.resolve(projectPath, '..')
    }
    const { id, is_uni_modules } = resolvePackage(filename)!
    const { code, msg } = await compilerServer.compile({
      projectPath,
      isCli,
      type: 1,
      pluginName: id,
      utsPath: resolveCompilerUtsPath(inputDir, is_uni_modules),
      swiftPath: resolveCompilerSwiftPath(outputDir, is_uni_modules),
    })
    result.code = code
    result.msg = msg
    result.changed = [swiftFile]
  }
  return result
}

function resolveCompilerUtsPath(projectPath: string, is_uni_modules: boolean) {
  return path.resolve(projectPath, is_uni_modules ? 'uni_modules' : 'utssdk')
}

function resolveCompilerSwiftPath(outputDir: string, is_uni_modules: boolean) {
  return path.resolve(outputDir, is_uni_modules ? 'uni_modules' : 'utssdk')
}

function isCliProject(projectPath: string) {
  if (projectPath.endsWith('src')) {
    return true
  }
  return false
}

export async function compile(
  filename: string,
  { inputDir, outputDir, sourceMap }: ToSwiftOptions
) {
  const { bundle, UtsTarget } = getUtsCompiler()
  // let time = Date.now()
  const result = await bundle(UtsTarget.SWIFT, {
    input: {
      root: inputDir,
      filename,
    },
    output: {
      isPlugin: true,
      outDir: outputDir,
      package: parseSwiftPackage(filename).namespace,
      sourceMap: sourceMap ? resolveUTSSourceMapPath() : false,
      extname: 'swift',
      imports: ['DCloudUTSFoundation'],
      logFilename: true,
      noColor: isInHBuilderX(),
    },
  })
  sourceMap &&
    moveRootIndexSourceMap(filename, {
      inputDir,
      outputDir,
      platform: 'app-ios',
      extname: '.swift',
    })
  return result
}

const deps = ['Info.plist', 'config.json']

export function resolveIOSDepFiles(filename: string) {
  const dir = resolveIOSDir(filename)
  return deps.map((dep) => path.resolve(dir, dep))
}

interface SwiftCompilerServer {
  compile(options: {
    projectPath: string
    isCli: boolean
    type: 1 | 2 // 1 => uni_modules , 2 => utssdk
    pluginName: string
    utsPath: string
    swiftPath: string
  }): Promise<{ code: number; msg: string }>
  checkEnv?: () => { code: number; msg: string }
}

export function checkIOSVersionTips(
  pluginId: string,
  pluginDir: string,
  is_uni_modules: boolean
) {
  const configJsonFile = join(
    pluginDir,
    is_uni_modules ? 'utssdk' : '',
    'app-ios',
    'config.json'
  )
  if (configJsonFile && fs.existsSync(configJsonFile)) {
    try {
      const configJson = parseJson(fs.readFileSync(configJsonFile, 'utf8'))
      if (configJson.deploymentTarget) {
        return `uts插件[${pluginId}]需在 iOS ${configJson.deploymentTarget} 版本及以上方可正常使用`
      }
    } catch (e) {}
  }
}
