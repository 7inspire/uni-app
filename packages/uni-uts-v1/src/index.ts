import { isArray } from '@vue/shared'

import { runKotlinProd, runKotlinDev } from './kotlin'
import { runSwiftProd, runSwiftDev } from './swift'

import { genProxyCode, resolvePlatformIndex, resolveRootIndex } from './code'
import { ERR_MSG_PLACEHOLDER, resolvePackage } from './utils'
import { parseUtsSwiftPluginStacktrace } from './stacktrace'
import { resolveUtsPluginSourceMapFile } from './sourceMap'
import { isWindows } from './shared'
import {
  generateCodeFrameWithKotlinStacktrace,
  generateCodeFrameWithSwiftStacktrace,
} from './legacy'

export const sourcemap = {
  generateCodeFrameWithKotlinStacktrace,
  generateCodeFrameWithSwiftStacktrace,
}

export * from './sourceMap'

export { compile as toKotlin } from './kotlin'
export { compile as toSwift } from './swift'

function parseErrMsg(code: string, errMsg: string) {
  return code.replace(ERR_MSG_PLACEHOLDER, errMsg)
}

function compileErrMsg(id: string) {
  return `uts插件[${id}]编译失败，无法使用`
}

export async function compile(module: string) {
  const pkg = resolvePackage(module)
  if (!pkg) {
    return
  }
  const deps: string[] = []
  const code = await genProxyCode(module, pkg)
  let errMsg = ''
  if (process.env.NODE_ENV !== 'development') {
    // 生产模式 支持同时生成 android 和 ios 的 uts 插件
    if (
      process.env.UNI_UTS_PLATFORM === 'app-android' ||
      process.env.UNI_UTS_PLATFORM === 'app'
    ) {
      const filename =
        resolvePlatformIndex('app-android', module, pkg) ||
        resolveRootIndex(module, pkg)
      if (filename) {
        await getCompiler('kotlin').runProd(filename)
      }
    }
    if (
      process.env.UNI_UTS_PLATFORM === 'app-ios' ||
      process.env.UNI_UTS_PLATFORM === 'app'
    ) {
      const filename =
        resolvePlatformIndex('app-ios', module, pkg) ||
        resolveRootIndex(module, pkg)
      if (filename) {
        await getCompiler('swift').runProd(filename)
      }
    }
  } else {
    // iOS windows 平台，标准基座不编译
    if (process.env.UNI_UTS_PLATFORM === 'app-ios') {
      if (isWindows) {
        process.env.UNI_UTS_TIPS = `iOS手机在windows上真机运行时uts插件代码修改需提交云端打包自定义基座才能生效`
        return {
          code,
          deps,
        }
      }
      if (process.env.HX_USE_BASE_TYPE === 'standard') {
        process.env.UNI_UTS_TIPS = `iOS手机在标准基座真机或模拟器运行暂不支持uts插件，如需调用uts插件请使用自定义基座`
        return {
          code,
          deps,
        }
      }
    }
    if (
      process.env.UNI_UTS_PLATFORM === 'app-android' ||
      process.env.UNI_UTS_PLATFORM === 'app-ios'
    ) {
      // dev 模式
      const filename =
        resolvePlatformIndex(process.env.UNI_UTS_PLATFORM, module, pkg) ||
        resolveRootIndex(module, pkg)
      const compilerType =
        process.env.UNI_UTS_PLATFORM === 'app-android' ? 'kotlin' : 'swift'

      if (filename) {
        deps.push(filename)
        const res = await getCompiler(compilerType).runDev(filename)
        if (res) {
          if (isArray(res.deps) && res.deps.length) {
            // 添加其他文件的依赖
            deps.push(...res.deps)
          }
          if (res.type === 'swift') {
            if (res.code) {
              errMsg = compileErrMsg(pkg.id)
              console.error(
                `error: ` +
                  (await parseUtsSwiftPluginStacktrace({
                    stacktrace: res.msg,
                    sourceMapFile: resolveUtsPluginSourceMapFile(
                      'swift',
                      filename,
                      process.env.UNI_INPUT_DIR,
                      process.env.UNI_OUTPUT_DIR
                    ),
                    sourceRoot: process.env.UNI_INPUT_DIR,
                  }))
              )
            }
          }
          const files: string[] = []
          if (process.env.UNI_APP_UTS_CHANGED_FILES) {
            try {
              files.push(...JSON.parse(process.env.UNI_APP_UTS_CHANGED_FILES))
            } catch (e) {}
          }
          if (res.changed && res.changed.length) {
            files.push(...res.changed)
          } else {
            if (res.type === 'kotlin') {
              errMsg = compileErrMsg(pkg.id)
            }
          }
          process.env.UNI_APP_UTS_CHANGED_FILES = JSON.stringify([
            ...new Set(files),
          ])
        } else {
          errMsg = compileErrMsg(pkg.id)
        }
      }
    }
  }
  return {
    code: parseErrMsg(code, errMsg),
    deps,
  }
}

function getCompiler(type: 'kotlin' | 'swift') {
  if (type === 'swift') {
    return {
      runProd: runSwiftProd,
      runDev: runSwiftDev,
    }
  }
  return {
    runProd: runKotlinProd,
    runDev: runKotlinDev,
  }
}
