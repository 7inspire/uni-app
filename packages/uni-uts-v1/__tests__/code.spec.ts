import { resolve } from 'path'
import { genProxyCode } from '../src/code'
import { ERR_MSG_PLACEHOLDER } from '../src/utils'

const pluginDir = resolve(__dirname, 'examples/uts/utssdk/test-uts')

describe('code', () => {
  test('genProxyCode', async () => {
    expect(
      (
        await genProxyCode(pluginDir, {
          id: 'test-uts',
          is_uni_modules: false,
          name: 'test-uts',
          namespace: 'uts.sdk.testUts',
          extname: '.uts',
        })
      ).replace(ERR_MSG_PLACEHOLDER, '')
    ).toMatchSnapshot()
  })
})
