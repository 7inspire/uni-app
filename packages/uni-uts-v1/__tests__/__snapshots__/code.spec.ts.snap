// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code genProxyCode 1`] = `
"
import { initUtsProxyClass, initUtsProxyFunction, initUtsPackageName, initUtsIndexClassName, initUtsClassName } from '@dcloudio/uni-app'
// const { initUtsProxyClass, initUtsProxyFunction, initUtsPackageName, initUtsIndexClassName, initUtsClassName } = uni
const name = 'test-uts'
const errMsg = \`\`
const is_uni_modules = false
const pkg = initUtsPackageName(name, is_uni_modules)
const cls = initUtsIndexClassName(name, is_uni_modules)

export const onMemoryWarning = initUtsProxyFunction(false, { errMsg, main: true, package: pkg, class: cls, name: 'onMemoryWarning', params: [{"name":"callback","type":"UTSCallback"}]})
export const offMemoryWarning = initUtsProxyFunction(false, { errMsg, main: true, package: pkg, class: cls, name: 'offMemoryWarning', params: [{"name":"callback","type":"UTSCallback","default":"UTSNull"}]})
"
`;
